Happy Dance/Assets/_Scripts/DanceDetector.cs

```diff
@@ -0,0 +1,95 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+
+using UnityEngine;
+using UnityEngine.Assertions;
+
+/// <summary>
+/// DanceDetector measures how much the player is moving their controller around and then plays a random praise track when it exceeds a certain value.
+/// DanceDetector then waits a random amount of time before being able to play another praise track.
+/// </summary>
+public class DanceDetector 
+    : MonoBehaviour
+{
+    public Transform HandTransform;
+    public float CommentActivationMovementThreshold;
+    public float TimeBeforeEncouragement;
+    [Space]
+    public List<Speech> PlayerPraiseOptions;
+    public List<Speech> PlayerEncouragementOptions;
+    public List<SpeechBubble> SpeechBubbles;
+    [Space]
+    public float BaseTalkCooldown;
+    public Vector2 CooldownRange;
+    public bool Active;
+
+    private Vector3 _cachedHandPos;
+    private Coroutine _talkRoutine;
+    private float _timeSinceGrossMovement;
+    private Speech _lastSpeechItem;
+
+    private void Start()
+    {
+        Assert.IsNotNull(HandTransform, "HandTransform must not be null!");
+        _cachedHandPos = HandTransform.position;
+        Active = true;
+    }
+
+    private void Update()
+    {
+        if (!Active)
+            return;
+
+        var handDelta = HandTransform.position - _cachedHandPos;
+
+        if (_talkRoutine == null)
+        {
+            _timeSinceGrossMovement += Time.deltaTime;
+
+            if (handDelta.magnitude > CommentActivationMovementThreshold)
+            {
+                _talkRoutine = StartCoroutine(TalkRoutine(PlayerPraiseOptions));
+                _timeSinceGrossMovement = 0;
+            }
+
+            if (_timeSinceGrossMovement > TimeBeforeEncouragement)
+            {
+                _talkRoutine = StartCoroutine(TalkRoutine(PlayerEncouragementOptions));
+                _timeSinceGrossMovement = 0;
+            }
+        }
+
+        _cachedHandPos = HandTransform.position;
```

Comment:
Move this to line 45 and later say `if (_talkRoutine == null) return;`

Reply:
Done